homeassistant/components/calendar/todoist.py

```diff
@@ -0,0 +1,488 @@
+    to the task on the Todoist website), 'end_time' (what time the event is
+    due), 'start_time' (what time this event was last updated), 'overdue' (is
+    the task past its due date?), 'priority' (1-4, how important the task is,
+    with 4 being the most important), and 'all_tasks' (all tasks in this
+    project, sorted by how important they are).
+
+    'offset_reached', 'location', and 'friendly_name' are defined by the
+    platform itself, but are not used by this component at all.
+
+    The 'update' method polls the Todoist API for new projects/tasks, as well
+    as any updates to current projects/tasks. This is throttled to every
+    MIN_TIME_BETWEEN_UPDATES minutes.
+    """
+
+    def __init__(self, project_data, labels, api,
+                 latest_task_due_date=None, whitelisted_labels=None,
+                 whitelisted_projects=None):
+        """Initialize a Todoist Project."""
+        self.event = None
+
+        self._api = api
+        self._name = project_data.get('name')
+        # If no ID is defined, fetch all tasks.
+        self._id = project_data.get('id')
+
+        # All labels the user has defined, for easy lookup.
+        self._labels = labels
+        # Not tracked: order, indent, comment_count.
+
+        self.all_project_tasks = []
+
+        # The latest date a task can be due (for making lists of everything
+        # due today, or everything due in the next week, for example).
+        if latest_task_due_date is not None:
+            self._latest_due_date = dt.utcnow() + timedelta(
+                days=latest_task_due_date)
+        else:
+            self._latest_due_date = None
+
+        # Only tasks with one of these labels will be included.
+        if whitelisted_labels is not None:
+            self._label_whitelist = whitelisted_labels
+        else:
+            self._label_whitelist = []
+
+        # This project includes only projects with these names.
+        if whitelisted_projects is not None:
+            self._project_id_whitelist = whitelisted_projects
+        else:
+            self._project_id_whitelist = []
+
+    def create_todoist_task(self, data):
+        """
+        Create a dictionary based on a Task passed from the Todoist API.
+
+        Will return 'None' if the task is to be filtered out.
+        """
+        task = {}
+        # Fields are required to be in all returned task objects.
+        task['summary'] = data['content']
+        task['completed'] = data['checked'] == 1
+        task['priority'] = data['priority']
+        task_url = 'https://todoist.com/showTask?id={}'.format(data['id'])
+        task['description'] = task_url
+
+        # All task Labels (optional parameter).
+        task['labels'] = [
+            label['name'] for label in self._labels
+            if label['id'] in data['labels']]
+
+        whitelist = self._label_whitelist
+        if any(label in task['labels'] for label in whitelist):
```

Comment:
I had tried that (you had suggested that change last time), and for whatever reason everything was inverted. I'm not exactly sure why it works, but it works.

Reply:
Looking at this again. Shouldn't all labels in `task[LABELS]` be on the whitelist to be ok? Otherwise we should return `None`. Then I think it should be:
```py
if not all(label in self._label_whitelist) for label in task[LABELS]:
    return None
```

Reply:
@MartinHjelmare 
So, here's the results with `if not all`:

![if not all](https://user-images.githubusercontent.com/2058763/30345070-a356d92a-97b8-11e7-87fa-cdc20a0f6648.png)

And here is `if any`:

![if any](https://user-images.githubusercontent.com/2058763/30345105-c3dc330c-97b8-11e7-8e8b-7cf8b6d534aa.png)

`if any` gives the correct results (with the label whitelist working as an `or` statement, [as documented here](https://github.com/home-assistant/home-assistant.github.io/pull/3283/files#diff-862da73f775be0e9540cb46d2513aa26R40)).

`if not all(label for self._label_whitelist) for label in task[LABELS]` causes *any* task with a label to be rejected for some reason. You can see that in the first screenshot -- the project "Mathematical Structures II only consists of tasks with the label "Homework" or "School" (or both), so `calendar.mathematical_structures_ii` rejects all of them. This change propagates down to my "custom" calendars specified in my configuration: `calendar.math_homework` and `calendar.due_today` are only taking into account tasks which don't have labels. You can see `calendar.movies_to_watch` is not affected, as none of the tasks there have labels.

The end result is only tasks with exactly 0 labels are accepted, with the whitelist not being taken into account at all.

`if any(label in task[LABELS] for label in self._label_whitelist)` returns the correct values, but I'm not sure why. I agree with you that there should be a `not` in there somewhere, but a `not` gives me inverted inputs.

I'll take a look at what's going on exactly and see if I can fix it. There must be some logic error somewhere. Maybe `if any(label in task[LABELS] for label in self._label_whitelist)` is just never returning true, and thus never tossing out anything?